src/zenml/model_deployers/base_model_deployer.py

```diff
@@ -0,0 +1,106 @@
+#  permissions and limitations under the License.
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional, Type
+
+from zenml.enums import ModelDeployerFlavor, StackComponentType
+from zenml.services import BaseService, ServiceConfig
+from zenml.stack import StackComponent
+
+
+class BaseModelDeployer(StackComponent, ABC):
+    """Base class for all ZenML model deployers."""
+
+    @property
+    def type(self) -> StackComponentType:
+        """The component type."""
+        return StackComponentType.MODEL_DEPLOYER
+
+    @property
+    @abstractmethod
+    def flavor(self) -> ModelDeployerFlavor:
+        """The model deployer flavor."""
+
+    @abstractmethod
+    def deploy_model(
+        self,    
+        config: ServiceConfig,
+    ) -> BaseService:
+        """Abstract method to deploy a model.
+
+        Concrete model deployer subclasses must implement the following
+        functionality in this method:
+        - Detect if there is an existing model server instance running serving
+        one or more previous versions of the same model
+        - Deploy the model to the serving platform or update the existing model
+        server instance to include the new model version
+        - Return a Service object that is a representation of the external model
+        server instance. The Service must implement basic operational state
+        tracking and lifecycle management operations for the model server (e.g.
+        start, stop, etc.)
+
+        Args:
+            config: Custom Service configuration parameters for the model
+                deployer. Can include the pipeline name, the run id, the step
+                name, the model name, the model uri, the model type etc. 
+
+        Returns:
+            The deployment Service object.
+        """
+
+    @abstractmethod
+    def find_model_server(
+        self,
+        pipeline_name: Optional[str] = None,
+        run_id: Optional[str] = None,
+        step_name: Optional[str] = None,
+        model_name: Optional[str] = None,
+        model_uri: Optional[str] = None,
+        model_type: Optional[str] = None,
+    ) -> List[BaseService]:
+        """Abstract method to find one or more a model servers that match the
+        given criteria.
+
+        Args:
+            pipeline_name: Name of the pipeline that the deployed model was part
+            of.
+            run_id: ID of the pipeline run which the deployed model was part of.
+            step_name: The name of the pipeline model deployment step that
+                deployed the model.
+            model_name: Name of the deployed model.
+            model_uri: URI of the deployed model.
+            model_type: Type/format of the deployed model.
+
+        Returns:
+            One or more Service objects representing model servers that match
+            the input search criteria.
+        """
+
+    @abstractmethod
+    def stop_model_server(self) -> None:
```

Comment:
This needs to accept the (service) UUID as a parameter. This is the only way to uniquely identify a service from both code and the CLI.